ecr-login/cli/docker-credential-ecr-login/main.go

```diff
@@ -24,5 +27,39 @@ import (
 func main() {
 	defer log.Flush()
 	config.SetupLogger()
-	credentials.Serve(ecr.ECRHelper{ClientFactory: api.DefaultClientFactory{}})
+	helper := ecr.ECRHelper{ClientFactory: api.DefaultClientFactory{}}
+	if len(os.Args) > 1 && os.Args[1] == "eval" {
+		evalCommand(helper)
+	} else {
+		credentials.Serve(helper)
+	}
+}
+
+func evalCommand(helper credentials.Helper) {
+	server, err := parseArgs(helper)
+	if err == nil {
+		var user, token string
+		user, token, err = helper.Get(server)
+		if err == nil {
+			fmt.Printf("docker login -u %s -p %s %s\n", user, token, server)
```

Comment:
I'm a little leery of doing this.  Docker has already changed the arguments that it accepts for `docker login`; it used to be that `--email` was required, but now providing `--email` fails.  You can see the challenge that we had with the AWS CLI and adapting it here: https://github.com/aws/aws-cli/issues/1926.

Reply:
I know; the latest version of docker is recommending to pass the arguments over stdin using `--password-stdin` flag. It is very annoying that the CLI is not kept backwards compatible.

My primary use case was running `eval $(docker run --rm -e AWS_REGION=us-east-1 dolbylabs/amazon-ecr-credential-helper)` in our normal EC2 instances on startup to be able to login to ECR. Since we pull our images only at startup, this was sufficient.

The only other alternative is to vendor the `moby/client` and call [`RegistryLogin()`](https://godoc.org/github.com/moby/moby/client#Client.RegistryLogin) function directly. I can make this change if you no not mind the extra vendor dependencies. I can still use the resulting image by bind mounting docker socket.

Reply:
I don't think that using `moby/client`/`RegistryLogin()` with the Docker socket will actually cause the `~/.docker/config.json` file to be written since the Docker socket controls the Docker daemon and not the client.

I think your use-case makes sense, but I'm still concerned as `docker login` does change (and the change you mentioned I didn't even know about until you said it here!).  Maybe instead of just calling it `eval`, we can encode the expected Docker CLI version in the name of the command somehow?  Something like `eval-1.11` so that it produces a `docker login` command without `--email`?

Reply:
I can change `eval` to `eval-1.11` so it is more explicit. I can also add `eval-17.06` with following code: 
  `fmt.Printf("echo %s | docker login --password-stdin -u %s %s", token, user, server)`

It will help to silence warnings from docker.

Reply:
The email flag for docker login has been optional since [docker v1.4](https://docs.docker.com/v1.4/reference/commandline/cli/#login). So I do not think we need to worry about the email flag. I can add `--password-stdin` flag to eval command to handle the new way and add `-e` flag without options to output email.

Reply:
I don't know offhand, but I'm pretty sure email was still required as of Docker 1.9.  I think I would prefer separate commands that correspond to the first version where the Docker CLI expected the invocation.